### ISSUE TITLE
Panic when completing `use itertools::`; structs with `where` cause ICE when completing

### ISSUE THREAD

**Issue by bluss** [racer-rust/racer#505]
- racer version 1.2.4 ab33b74 (the tag)
- crate version itertools 0.4.7
- rust version rustc 1.8.0-nightly (34af2de40 2016-02-05)

racer can complete with a more restricted match, for example completing `use itertools::Z` is fine. It panics when completing `use itertools::`

```
$ racer complete 77 15 src/lib.rs
PREFIX 15,15,
error: internal compiler error: Error constructed but not emitted
thread 'searcher' panicked at 'explicit panic', .cargo/registry/src/github.com-88ac128001ac3a9a/syntex_syntax-0.26.0/src/errors/mod.rs:261
note: Run with `RUST_BACKTRACE=1` for a backtrace.
ERROR:racer: Search thread paniced: Any
END
```

It actually hangs when trying to get the backtrace:

```
$ RUST_BACKTRACE=1 racer complete 77 15 src/lib.rs
PREFIX 15,15,
error: internal compiler error: Error constructed but not emitted
thread 'searcher' panicked at 'explicit panic', .cargo/registry/src/github.com-88ac128001ac3a9a/syntex_syntax-0.26.0/src/errors/mod.rs:261
stack backtrace:
   1:     0x558353f663c0 - sys::backtrace::tracing::imp::write::h5818e2087d5d7662Squ
   2:     0x558353f689bb - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43119
   3:     0x558353f68628 - panicking::default_handler::h088b3b337e972033xQy
   4:     0x558353f5832c - sys_common::unwind::begin_unwind_inner::hc933f5a4d9faf478qjt
   5:     0x558353dc7004 - sys_common::unwind::begin_unwind::h6935125483258856894
^C
```

**Comment by bluss** [comment]
I hit submit too early. Editing!

**Comment by bluss** [comment]
There exists both `ìtertools::Coalesce` and `itertools::Combinations`, `itertools::CombinationsN`.

These completions fail (panic)

- `itertools::C`
- `itertools::Co`
- `itertools::Coa`  (this one should be unique)

These completions succeed

- `itertools::Com`

So `Coalesce` for some reason is one of several items that won't complete, and if it would match, it panics the completion.

**Comment by bluss** [comment]
I found it! The common factor is that `struct`s with `where` clauses will panic the completion, and the structs that are using bounds inlined in the declaration will succeed.

Panics completion:

```rust
pub struct Coalesce<I, F>
    where I: Iterator,
{
    iter: CoalesceCore<I>,
    f: F,
}
```

Succeeds completion:

```rust
pub struct Coalesce<I: Iterator, F>
{
    iter: CoalesceCore<I>,
    f: F,
}
```

**Comment by phildawes** [comment]
@bluss thanks for this analysis; If I get a chance tonight I'll try to fix this

**Comment by bluss** [comment]
FYI using racer 1.2.5 / syntex 0.24.0, it looks like the `where`-structs can be completed again.

However there is something else that causes completion after `use itertools::` to overflow the stack. Seems to be related to `itertools::linspace::linspace` (where `linspace` is an internal module..) :frowning: which is a bug that's already been reported — #354

**Comment by letheed** [comment]
Can't reproduce using itertools-0.4.7 with racer-2.0.9.

Should probably be closed since #354 was closed as solved.

**Comment by jwilm** [comment]
Thanks @letheed
